{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOd8r7UsXW0Be6bx3PrAgMK",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/kanvaudupa-bioinfo/Bioinfo-lab/blob/main/Info_Prog5_Restriction_mapper.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0ce46c5b"
      },
      "source": [
        "Now, let's write a function to read the FASTA file. This function will parse the file and return a dictionary where keys are sequence headers and values are the corresponding DNA sequences."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5b20f121"
      },
      "source": [
        "## Upload your FASTA file\n",
        "Please upload your FASTA file using the button below. This file will be used for the restriction site analysis."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 110
        },
        "id": "ab4dcdc8",
        "outputId": "3141bbc4-d219-43fe-d8e8-6c7fcfa59d15"
      },
      "source": [
        "from google.colab import files\n",
        "import os\n",
        "\n",
        "print(\"Please upload your FASTA file (e.g., genome.fasta).\")\n",
        "uploaded = files.upload()\n",
        "\n",
        "# Get the name of the first uploaded file\n",
        "if uploaded:\n",
        "    global uploaded_fasta_file_name\n",
        "    uploaded_fasta_file_name = next(iter(uploaded.keys())) # Use .keys() to get the filename\n",
        "    print(f\"File '{uploaded_fasta_file_name}' uploaded successfully.\")\n",
        "else:\n",
        "    uploaded_fasta_file_name = None\n",
        "    print(\"No file was uploaded. Please upload a FASTA file to proceed.\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Please upload your FASTA file (e.g., genome.fasta).\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-ea1eca3d-faea-4f3b-a01e-e1d1833461df\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-ea1eca3d-faea-4f3b-a01e-e1d1833461df\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving APP.fasta to APP.fasta\n",
            "File 'APP.fasta' uploaded successfully.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "842c005b",
        "outputId": "c4556c11-428d-497c-8eff-3e0b303d060b"
      },
      "source": [
        "def read_fasta(file_path):\n",
        "    sequences = {}\n",
        "    current_sequence_name = None\n",
        "    with open(file_path, 'r') as f:\n",
        "        for line in f:\n",
        "            line = line.strip()\n",
        "            if not line:\n",
        "                continue\n",
        "            if line.startswith('>'):\n",
        "                current_sequence_name = line[1:].split()[0] # Get header, strip '>' and take first word\n",
        "                sequences[current_sequence_name] = []\n",
        "            elif current_sequence_name:\n",
        "                sequences[current_sequence_name].append(line.upper()) # Convert to uppercase for consistent matching\n",
        "\n",
        "    # Join list of sequence lines into a single string\n",
        "    for name, seq_list in sequences.items():\n",
        "        sequences[name] = ''.join(seq_list)\n",
        "    return sequences\n",
        "\n",
        "print(\"FASTA reader function defined.\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "FASTA reader function defined.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dad91d06"
      },
      "source": [
        "Next, we'll define a function that takes the genome sequences and a list of restriction sites, then counts the occurrences of each site."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9a4332ce",
        "outputId": "961eaaca-1b23-46c1-801c-615dd4f10c7b"
      },
      "source": [
        "def count_restriction_sites(genome_sequences, restriction_sites):\n",
        "    # Initialize a dictionary to store lists of occurrence indices for each site\n",
        "    site_occurrences_locations = {site: [] for site in restriction_sites}\n",
        "\n",
        "    for seq_name, sequence in genome_sequences.items():\n",
        "        print(f\"Processing {seq_name}...\")\n",
        "        for site in restriction_sites:\n",
        "            i = 0\n",
        "            while True:\n",
        "                idx = sequence.find(site.upper(), i)\n",
        "                if idx == -1:\n",
        "                    break\n",
        "                # Store the 0-based index of the start of the match\n",
        "                site_occurrences_locations[site].append(idx)\n",
        "                i = idx + 1 # Move past the found site to find next (non-overlapping)\n",
        "\n",
        "    return site_occurrences_locations\n",
        "\n",
        "print(\"Restriction site counter function defined to also record locations.\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Restriction site counter function defined to also record locations.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "67a4739d",
        "outputId": "c942ef33-9e8e-4a4d-9acd-4801d01a5a11"
      },
      "source": [
        "# Define the path to your FASTA file\n",
        "# This will now use the file uploaded in the previous step\n",
        "if 'uploaded_fasta_file_name' in globals() and uploaded_fasta_file_name:\n",
        "    fasta_file = uploaded_fasta_file_name\n",
        "else:\n",
        "    print(\"Error: No FASTA file was uploaded or its name is not available.\")\n",
        "    print(\"Please ensure you've run the upload cell and successfully selected a file.\")\n",
        "    fasta_file = None # Set to None to prevent further execution if no file\n",
        "\n",
        "if fasta_file:\n",
        "    # Read the genome sequences\n",
        "    genome_data = read_fasta(fasta_file)\n",
        "\n",
        "    # Define the restriction enzymes and their sites as a dictionary\n",
        "    # Key: Restriction enzyme name, Value: Restriction sequence\n",
        "    restriction_enzymes = {\n",
        "        'EcoRI': 'GAATTC',\n",
        "        'HindIII': 'AAGCTT',\n",
        "        'BamHI': 'GGATCC',\n",
        "        'PvuII': 'CAGCTG',\n",
        "        'SmaI': 'CCCGGG'\n",
        "    }\n",
        "\n",
        "    # Extract just the sequences to pass to the counting function\n",
        "    restriction_sequences = list(restriction_enzymes.values())\n",
        "\n",
        "    # Count the occurrences of each restriction site and get their locations\n",
        "    all_occurrences_with_locations = count_restriction_sites(genome_data, restriction_sequences)\n",
        "\n",
        "    # Display the results, mapping back to enzyme names\n",
        "    print(\"\\nRestriction enzyme, site, and cut locations:\")\n",
        "    for enzyme_name, sequence in restriction_enzymes.items():\n",
        "        # Retrieve the list of locations for the current site\n",
        "        locations = all_occurrences_with_locations.get(sequence.upper(), [])\n",
        "        count = len(locations)\n",
        "\n",
        "        print(f\"Enzyme '{enzyme_name}' (Site: '{sequence}'):\")\n",
        "        print(f\"  Total occurrences: {count}\")\n",
        "        if locations:\n",
        "            # Display locations as a comma-separated list, adding 1 for 1-based indexing\n",
        "            print(f\"  Cut locations (1-indexed start): {', '.join(map(str, [loc + 1 for loc in locations]))}\")\n",
        "        else:\n",
        "            print(f\"  No cut locations found.\")\n",
        "else:\n",
        "    print(\"Skipping analysis due to missing FASTA file.\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Processing AF293341.1...\n",
            "\n",
            "Restriction enzyme, site, and cut locations:\n",
            "Enzyme 'EcoRI' (Site: 'GAATTC'):\n",
            "  Total occurrences: 1\n",
            "  Cut locations (1-indexed start): 1286\n",
            "Enzyme 'HindIII' (Site: 'AAGCTT'):\n",
            "  Total occurrences: 0\n",
            "  No cut locations found.\n",
            "Enzyme 'BamHI' (Site: 'GGATCC'):\n",
            "  Total occurrences: 0\n",
            "  No cut locations found.\n",
            "Enzyme 'PvuII' (Site: 'CAGCTG'):\n",
            "  Total occurrences: 1\n",
            "  Cut locations (1-indexed start): 408\n",
            "Enzyme 'SmaI' (Site: 'CCCGGG'):\n",
            "  Total occurrences: 1\n",
            "  Cut locations (1-indexed start): 1318\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2f9e0ec8"
      },
      "source": [
        "## Primer Design\n",
        "Now, let's design primers that are complementary to the sequence approximately 20 base pairs downstream (after) each identified restriction cut site. We'll generate the reverse complement of this 20bp sequence to serve as a primer."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "317bb0e6",
        "outputId": "c863ddd2-a413-44b2-94dd-b79fbea3acaf"
      },
      "source": [
        "def reverse_complement(dna_sequence):\n",
        "    complement_map = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'}\n",
        "    complement_sequence = ''.join([complement_map[base] for base in dna_sequence.upper()])\n",
        "    return complement_sequence[::-1]\n",
        "\n",
        "primer_length = 20\n",
        "\n",
        "print(f\"Designing primers (approx. {primer_length}bp) for sequences immediately after each restriction cut site (1-indexed cut position)...\")\n",
        "\n",
        "# Assuming genome_data, restriction_enzymes, and all_occurrences_with_locations are available from previous cells\n",
        "\n",
        "for enzyme_name, site_sequence in restriction_enzymes.items():\n",
        "    # Get locations for the current site (using the uppercase site sequence as key)\n",
        "    # The locations are 0-indexed where the site starts\n",
        "    locations_0_indexed = all_occurrences_with_locations.get(site_sequence.upper(), [])\n",
        "\n",
        "    if locations_0_indexed:\n",
        "        print(f\"\\nEnzyme '{enzyme_name}' (Site: '{site_sequence}'):\")\n",
        "        for loc_0_indexed in locations_0_indexed:\n",
        "            # The cut is AT THE END of the restriction site for many enzymes.\n",
        "            # For simplicity, we are taking sequence *after* the restriction site sequence itself.\n",
        "            # If the cut is internal to the site, adjust `start_extract_index` accordingly.\n",
        "\n",
        "            # Start extracting immediately after the restriction site sequence\n",
        "            start_extract_index = loc_0_indexed + len(site_sequence)\n",
        "\n",
        "            # Assuming single sequence in genome_data for simplicity or pick the first one\n",
        "            # If there are multiple sequences in genome_data, you'd need to associate locations with specific sequences.\n",
        "            # For now, let's assume we are interested in the first sequence in genome_data values.\n",
        "            # This part needs refinement if FASTA has multiple sequences and primers are needed for each.\n",
        "\n",
        "            # For this example, let's iterate through all sequences in genome_data\n",
        "            for seq_header, genome_seq in genome_data.items():\n",
        "                if start_extract_index < len(genome_seq):\n",
        "                    # Extract sequence for primer design\n",
        "                    primer_template_sequence = genome_seq[start_extract_index : start_extract_index + primer_length]\n",
        "\n",
        "                    # Generate reverse complement\n",
        "                    designed_primer = reverse_complement(primer_template_sequence)\n",
        "\n",
        "                    # Print the results (1-indexed cut position for user clarity)\n",
        "                    print(f\"  Cut at {loc_0_indexed + 1}: Template '{primer_template_sequence}' -> Primer '{designed_primer}' (complementary to sequence starting at {start_extract_index + 1} in {seq_header})\")\n",
        "                else:\n",
        "                    print(f\"  Cut at {loc_0_indexed + 1}: Not enough sequence ({primer_length}bp) after cut site in {seq_header} to design primer.\")\n",
        "    else:\n",
        "        print(f\"\\nEnzyme '{enzyme_name}' (Site: '{site_sequence}'): No cut locations found for primer design.\")\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Designing primers (approx. 20bp) for sequences immediately after each restriction cut site (1-indexed cut position)...\n",
            "\n",
            "Enzyme 'EcoRI' (Site: 'GAATTC'):\n",
            "  Cut at 1286: Template 'CAGGGATGAATGGGCAAAAG' -> Primer 'CTTTTGCCCATTCATCCCTG' (complementary to sequence starting at 1292 in AF293341.1)\n",
            "\n",
            "Enzyme 'HindIII' (Site: 'AAGCTT'): No cut locations found for primer design.\n",
            "\n",
            "Enzyme 'BamHI' (Site: 'GGATCC'): No cut locations found for primer design.\n",
            "\n",
            "Enzyme 'PvuII' (Site: 'CAGCTG'):\n",
            "  Cut at 408: Template 'GGAGGAGAGAAGAAAGCGGG' -> Primer 'CCCGCTTTCTTCTCTCCTCC' (complementary to sequence starting at 414 in AF293341.1)\n",
            "\n",
            "Enzyme 'SmaI' (Site: 'CCCGGG'):\n",
            "  Cut at 1318: Template 'TTGCCTGGAGCAGTAGGACA' -> Primer 'TGTCCTACTGCTCCAGGCAA' (complementary to sequence starting at 1324 in AF293341.1)\n"
          ]
        }
      ]
    }
  ]
}
